### PR TITLE
sql: adjust ulp errors for arbitrary base logs w/ numerics

### PR DESCRIPTION
**Pull request by sploiselle**
### Motivation

This PR fixes a recognized bug. I previously ascribed this limitation as intrinsic to the library, but in the patched situation (`log(numeric, numeric)`), the inherent imprecision in dividing irrational numbers was another factor in this behavior. However, very possible to improve the results in the narrow case where the "ulp error" crops up adjacent to an integer.

### Description

The code manually adjusts results whose values are only "ulp" away form their nearest integral, and uses this function in the context of arbitrary base logs. This fixes the most common complaint from users by turning

```sql
SELECT log(2::numeric, 64)
----
5.99999999999999999999999999999999999999
```

into 
```sql
SELECT log(2::numeric, 64)
----
6
```

### Checklist

- [x] This PR has adequate test coverage.
- [x] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

**Comment by benesch** [comment]
So, sorry Sean, but I'm super confused about the approach here! I was expecting something like the following, which just rounds to one less digit of precision:

```diff
diff --git a/src/expr/src/scalar/func.rs b/src/expr/src/scalar/func.rs
index 7003c757b..093401797 100644
--- a/src/expr/src/scalar/func.rs
+++ b/src/expr/src/scalar/func.rs
@@ -1532,7 +1532,11 @@ fn log_base_numeric<'a>(a: Datum<'a>, b: Datum<'a>) -> Result<Datum<'a>, EvalErr
     if a.is_zero() {
         Err(EvalError::DivisionByZero)
     } else {
-        let mut b = numeric::adjust_ulp(b);
+        let mut cx = numeric::cx_datum();
+        unsafe {
+            (*(&mut cx as *mut dec::Context<Numeric> as *mut decnumber_sys::decContext)).digits -= 1;
+        }
+        cx.reduce(&mut b);
         numeric::munge_numeric(&mut b).unwrap();
         Ok(Datum::from(b))
```

That has the effect of rounding `5.99999999999999999999999999999999999999` to `6` as desired. The gross bit of using `unsafe` to decrement `context.digits` would require a safer API, of course.

It's late, I'm tired, and I'm afraid I'm struggling to put into words exactly what I find confusing about the current approach, so sorry if this comes out a bit garbled. I'm pretty sure `adjust_ulp` is roughly equivalent to the above code, except that it uses what I think is a bespoke rounding mode whose semantics I'm struggling to describe. It's not quite any of the existing rounding modes ("half up", "away from zero", etc.), I don't think. I also think the approach is designed around these ULP errors when I'm not convinced we've seen evidence of such errors. Specifically, here's what decnumber computes ln(2) and ln(64) as compared to what WolframAlpha says:

```
Materialize:  ln(2) = 0.693147180559945309417232121458176568076
WolframAlpha: ln(2) = 0.6931471805599453094172321214581765680755001343602552541206800094

Materialize:  ln(64) = 4.15888308335967185650339272874905940845
WolframAlpha: ln(64) = 4.1588830833596718565033927287490594084530008061615315247240800569
```

In both cases those logarithms look correctly rounded to me, so I don't think that rare 1 ULP error is occurring; I think it's just that dividing approximations results in an inexact result.

Finally, based on my read of the decnumber docs, the error that occurs seems like it causes the result to be off by _1_ ULP. I think that means the last digit might be +/- 1 from what it _should_ be, not that the last digit is entirely wrong?

**Comment by sploiselle** [comment]
ty for the review, @benesch.

I think the struggle to see the semantics of my code is due to an oversight: it
failed to account for which "direction" the final digit should round. Could be
fixed with an additional check on the coefficient value generated from `diff`.
If I made that adjustment, the semantics of the two approaches are basically
equivalent or can be made so.

However, I didn't "see" your proposed solution as a possibility because it
relies on a side-effect of calling `reduce` to perform the rounding, i.e. the
call winds up invoking `decSetCoeff` in the C `decnumber` library, which
actually performs the rounding of this final digit. Clever! How did you find
that?

I also want to be clear that I was trying to state that this behavior _was not_
caused by one of the library's 1-ULP errors on log. I was borrowed terminology
in a confusing context, and I should have instead said "least-significant
digit," rather than "ulp." I'll evidence by the fact that I wasn't correcting
for 1-ulp errors, but instead "arbitrary ulp errors".

I'm working on extending `rust-dec` to let users change the digits on `Context`
and will send that PR up in a bit.

**Comment by sploiselle** [comment]
Ok––this is good for review again; had forgotten about `set_precision` which makes this trivial.

File: src/expr/Cargo.toml [hunk 1]
```diff
@@ -11,6 +11,7 @@ anyhow = "1.0.44"
 chrono = { version = "0.4.0", default-features = false, features = ["clock", "std"] }
 csv = "1.1.6"
 dec = "0.4.5"
+decnumber-sys = "0.1.5"
```
Comment (sploiselle) [inline src/expr/Cargo.toml:14]: Derp; will delete this

**Review by benesch — APPROVED** [review]
Looooool we already had a function to adjust the precision? 🤦🏽‍♂️ Well this has been an absolute success story of tag teaming from my perspective, hope you agree!

**Pull request merged**